File: Grind75_1/week2/ReverseLinkedList.py
"""
https://leetcode.com/problems/reverse-linked-list/description/
"""

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:

    def reverseListItr(self, head: Optional[ListNode]) -> Optional[ListNode]:
        
        # use 3 pointers, 1 to keep track the target to reverse, 1 to keep track the prev node, and 1 to keep track of the next node.
        headRev = None

        while (head is not None):
            tempP = head.next
            head.next = headRev
            headRev = head
            head = tempP
            
        return headRev
        

    def reverseListRec(self, head: Optional[ListNode], newHead: Optional[ListNode]) -> Optional[ListNode]:

        if (head is None):
            return newHead

        tempP = head.next
        head.next = newHead

        return self.reverseListRec(tempP, head)


    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        #return self.reverseListItr(head)
        return self.reverseListRec(head, None)

File: neetcode/neet150_JS/Linked-List/reverse-linked-list.js
// https://neetcode.io/problems/reverse-a-linked-list

/**
 * Definition for singly-linked list.
 * class ListNode {
 *     constructor(val = 0, next = null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

class Solution {
    /**
     * @param {ListNode} head
     * @return {ListNode}
     */
    reverseList(head) {
        if (head === null) {
            return head
        }

        let prev = null
        let next = null

        while (head !== null) {
            next = head.next
            head.next = prev
            prev = head
            head = next
        }

        return prev
    }
}
